### PR TITLE
error handling

### PR DESCRIPTION
**Pull request by Pavel-Durov**
- error propagation and handling at the top level

>  I tried the anyhow crate, but then I decided to implement error handling myself to experience the problems it tries to address.

### CODE REVIEW

**Comment by vext01** [comment]
I'll review this after it has been rebased.

(It's probably best to hold off raising unrebased PRs? What do you think @ltratt?)

**Comment by Pavel-Durov** [comment]
@vext01 rebased and ready for review

File: README.md [hunk 1]
```diff
@@ -1,13 +1,10 @@
-[![Bors enabled](https://bors.tech/images/badge_small.svg)](https://app.bors.tech/repositories/61986)
-
 # pavel.yaiwr
 
-Yet Another Interpreter Written In Rust
+Yet Another Interpreter Witten In Rust
```
Comment (vext01) [inline README.md:3]: Are you sure the rebase was correct? Didn't this change appear in a prior PR?

Comment (Pavel-Durov) [reply README.md:3]: that's weird, will investigate :detective:

**Comment by vext01** [comment]
This looks fine!

bors r+

**Comment by Pavel-Durov** [comment]
this is interesting, I get different parsing errors when I run `cargo test --release` compared to `cargo test`.

**Comment by Pavel-Durov** [comment]
I've removed parsing error assertion from tests to make the build pass :point_right: 716fd550f99411f30ca68d1dc0a6e354361574ff
not sure if its something todo with the way `lrpar` is implemented or its something todo with my configurations

**Comment by Pavel-Durov** [comment]
bors try

**Comment by Pavel-Durov** [comment]
bors try

**Pull request merged**